File: Data Structures & Algorithms/validate-parentheses/submission-4.java
class Solution {
    public boolean isValid(String s) {
        Deque<Character> st = new ArrayDeque<>();

        for(char c: s.toCharArray()){
            // push all open brackets
            if(c == '(' || c == '[' || c == '{'){
                st.push(c);
            }else{
                // close bracket: stack must be non-empty and top must match
                if(st.isEmpty()) return false;
                
                char top = st.pop();
                if (c == ')' && top != '(') return false;
                if (c == ']' && top != '[') return false;
                if (c == '}' && top != '{') return false;
            }
        }
        return st.isEmpty();
    }
}

File: Data Structures & Algorithms/simplify-path/submission-4.py
class Solution:
    def simplifyPath(self, path: str) -> str:
        stack = []
        for p in path.split('/'):
            if p == "." or p == "":
                continue
            if p != "..":
                stack.append(p)
            elif stack:
                #  .. condition : pop when stack is not empty
                stack.pop()
        return '/' + '/'.join(stack)